src/client/render2d/pixi/rendering.ts

```diff
@@ -0,0 +1,110 @@
+import { ArcGeometry } from '../geometry/arc_geometry'
+import { ArrowGeometry } from '../geometry/arrow_geometry'
+import { CircleGeometry } from '../geometry/circle_geometry'
+import { LineGeometry } from '../geometry/line_geometry'
+import { MarkerGeometry } from '../geometry/marker_geometry'
+import { PathGeometry } from '../geometry/path_geometry'
+import { PolygonGeometry } from '../geometry/polygon_geometry'
+import { TextGeometry } from '../geometry/text_geometry'
+import { Group } from '../object/group'
+import { Object2d } from '../object/object2d'
+import { Shape } from '../object/shape'
+
+import { renderArc } from './arc'
+import { renderArrow } from './arrow'
+import { renderCircle } from './circle'
+import { renderLine } from './line'
+import { renderMarker } from './marker'
+import { renderPath } from './path'
+import { renderPolygon } from './polygon'
+import { renderText } from './text'
+
+export const pixiObject = createTransformer((obj: Object2d): DisplayObject => {
+
+  if (obj instanceof Group) {
+    const g = new Container()
+
+    const { transform: { scale, translate, rotate } } = obj
+    g.scale.x = scale.x
+    g.scale.y = scale.x
+    g.x = translate.x
+    g.y = translate.y
+    g.rotation = rotate
+
+    obj.children.forEach(o => {
+      g.addChild(pixiObject(o))
+    })
+
+    return g
+
+  } else if (obj instanceof Shape) {
+    if (obj.geometry instanceof ArcGeometry) {
+      return renderArc(obj)
+    } else if (obj.geometry instanceof ArrowGeometry) {
+      return renderArrow(obj)
+    } else if (obj.geometry instanceof CircleGeometry) {
+      return renderCircle(obj)
+    } else if (obj.geometry instanceof LineGeometry) {
+      return renderLine(obj)
+    } else if (obj.geometry instanceof MarkerGeometry) {
+      return renderMarker(obj)
+    } else if (obj.geometry instanceof PathGeometry) {
+      return renderPath(obj)
+    } else if (obj.geometry instanceof PolygonGeometry) {
+      return renderPolygon(obj)
+    } else if (obj.geometry instanceof TextGeometry) {
+      return renderText(obj)
+    } else {
+      throw new Error(`Unsupported geometry type: ${obj.geometry}`)
+    }
+  } else {
+    throw new Error(`Unsupported Object2d type: ${obj}`)
+  }
+})
+
+const toPixiColor = (style: string): {color: number, alpha: number} => {
+  if (style === 'transparent') {
+    return { color: 0, alpha: 0 }
+  }
+  // Colors of the form #FFF
+  let result = /^#([A-Fa-f0-9])([A-Fa-f0-9])([A-Fa-f0-9])$/.exec(style)
+  if (result !== null) {
+    return{ color: parseInt(result[1] + result[1] + result[2] + result[2] + result[3] + result[3], 16), alpha: 1 }
+  }
+  // Colors of the form #FFFFFF
+  result = /^#([A-Fa-f0-9]{6})$/.exec(style)
+  if (result !== null) {
+    return { color: parseInt(result[1], 16), alpha: 1 }
+  }
+
+  throw new Error('Pixi cannot handle non hex colours')
+}
+
+export function applyAppearance(obj: Graphics, appearance: Appearance, draw: (obj: Graphics) => void): void {
```

Comment:
I'd like to explore how we can remove the need for the `draw` callback, but that is for another PR I think.

Reply:
The draw callback is used to wrap the drawing code with an appearance.
It handles code that needs to be executed before/after drawing to make it look how we want.